python/tskit/ms.py

```diff
@@ -0,0 +1,129 @@
+        self._tree_sequence = tree_sequence
+        self._num_replicates = num_replicates
+        self._mutation_rate = mutation_rate
+        self._precision = precision
+        self._print_trees = print_trees
+        self._num_loci = num_loci
+
+    def print_trees(self, output):
+        """
+        Print out the trees in ms-format from the specified tree sequence.
+        When 'invisible' recombinations occur ms prints out copies of the
+        same tree. Therefore, we must keep track of all breakpoints from the
+        simulation and write out a tree for each one.
+        """
+        breakpoints = list(self._tree_sequence.breakpoints(True)) + [self._num_loci]
+        if self._num_loci == 1:
+            tree = next(self._tree_sequence.trees())
+            newick = tree.newick(precision=self._precision)
+            print(newick, file=output)
+        else:
+            j = 1
+            for tree in self._tree_sequence.trees():
+                newick = tree.newick(precision=self._precision)
+                left, right = tree.interval
+                while j < (len(breakpoints) - 1) and breakpoints[j] <= right:
+                    length = breakpoints[j] - left
+                    left = breakpoints[j]
+                    j += 1
+                    # Print these seperately to avoid the cost of creating
+                    # another string.
+                    print("[{}]".format(int(length)), end="", file=output)
+                    print(newick, file=output)
+
+    def __write_header(self, output):
+        print(
+            "ms {} {} # This file is an ms-style output file generated from tskit. The two arguments written are sample size and number of replicates".format(
+                self._tree_sequence.sample_size, max(self._num_replicates, 1)
+            ),
+            file=output
+        )
+        print("{} # Setting random seed to 999".format(999), file=output)
+
+    def write(self, output):
+
+        if os.path.getsize(output.fileno()) == 0:
+            self.__write_header(output)
+
+        tree_sequence = self._tree_sequence
+        print(file=output)
+        print("//", file=output)
+        if self._print_trees:
+            self.print_trees(output)
+        if self._mutation_rate > 0:
+            s = tree_sequence.get_num_sites() 
+            print("segsites:", s, file=output)
+            if s != 0:
+                print("positions: ", end="", file=output)
+                positions = [
+                    variant.position / (tree_sequence.sequence_length)
+                    for variant in tree_sequence.variants()
+                ]
+                positions.sort()
+                for position in positions:
+                    print(
+                        "{0:.{1}f}".format(position, self._precision),
+                        end=" ",
+                        file=output,
+                    )
+                print(file=output)
+                for h in tree_sequence.haplotypes():
```

Comment:
Hm - so, `haplotypes` returns the actual alleles, which are thus required to be 0/1. But the `genotype_matrix` gives the indexes into arrays of alleles, with `0` always the ancestral state. So, more generally, could we do something like this:
```
genotypes = tree_sequence.genotype_matrix()
for k in range(tree_sequence.num_samples):
  print("".join(genotypes[:, k]), file=output)
```
(but, what's the `-` for?)

Reply:
(The "-" is for missing data @petrelharp )

Reply:
that's what I guessed... but does ms output missing data?

Reply:
No, I wouldn't imagine it does. But, I'm not sure how strict we should be here?

Reply:
I'm not opposed to including missing data! Just wanted to be sure what was happening.